Update tools/essence-feature-usage-stats

gskorokhod opened the pull request:
# Done

- [x] Update to work with the new directory structure in conjure v2.5.1
- [x] Specify multiple repos for Essence files
- [x] Make repos sortable / hideable
- [x] Add environment variable to cap the number of Essence files that get processed
- [x] Add a way to exclude paths that match a given regex (By default, exclude all autogen files)
- [x] Switched to data tables plugin for the table sorting features
- [x] Added exporting data as CSV

This tool should be done and ready for review now!

ozgurakgun commented:
@gskorokhod merged main one more time to bring in the new github actions changes, will stop messing up your PR now :)

gskorokhod commented:
Good afternoon! I have fixed the sorting issues by switching to the DataTables plugin instead of my own sorting code, and added a link to download data as a CSV file. There is no Python lint errors and I've documented the new env variables in the project's README so should be ready to merge! 
(Checks fail because it says it doesn't have permissions to deploy - The page builds fine in my fork so i assume it will be fine once this is merged?)

niklasdewally commented:
> (Checks fail because it says it doesn't have permissions to deploy - The page builds fine in my fork so i assume it will be fine once this is merged?)

This doesn't seem related to your changes btw.

---

**Ping**: @ozgurakgun

Workflows don't get write access on PRs for security.

```
on:
  push:
    branches:
      - main # run for pushes to the main branch
  workflow_dispatch:
  pull_request:
    paths:
      - tools/essence-feature-usage-stats/**
```

I would remove PR from the above in the CI file.


Alternatively, if you want this to run in PRs successfully, you can use a workflow_run rule to do this.

In short, workflow A runs in the context of the PR, generates all the pages, and stores them as artifacts.
Workflow B runs in the context of main, and is triggered when workflow A finishes. It loads the artifact saved by A, and deploys it to GH pages. As workflow B runs on main, it has write permissions.

See code-coverage / code-coverage-deploy for existing examples of this.

niklasdewally commented:
The above basically ensures that a malicious contributor does not edit CI in a PR, and that edited CI nukes the repo.
Workflow B always runs the copy of itself on main, so is safe.

ozgurakgun commented:
We still want to build for PRs, but not push/deploy. An if statement to the deploy step should solve it?

gskorokhod commented:
> We still want to build for PRs, but not push/deploy. An if statement to the deploy step should solve it?

I will have a look!

gskorokhod commented:
Added the if statement
By the way, why do we want to build on pull requests if it is not deployed? Is it just to ensure that it still builds correctly?

ozgurakgun commented:
> Added the if statement By the way, why do we want to build on pull requests if it is not deployed? Is it just to ensure that it still builds correctly?

exactly!

ozgurakgun commented on .github/workflows/essence-feature-stats.yml:
```diff
@@ -12,9 +12,13 @@ on:
 env:
   ESSENCE_DIR: "./EssenceCatalog"
```
does this still need to be in the ENV? did you do multiple repos yet in this PR?

gskorokhod replied:
Yes! Repos are specified in another variable (ESSENCE_FILE_REPOS I think?)
This is just the root directory for essence files, all repos will be cloned to its sub directories. I think it doesn’t matter for GitHub actions but this variable is still expected to be set

ozgurakgun replied:
is it not a bit confusing to have it set to EssenceCatalog, which is one of the repo names?

gskorokhod replied:
true, changed that

gskorokhod commented:
I'm rearranging my branches (making main an exact copy of the main in conjure-cp/conjure-oxide, and creating dev branches for every feature I'm working on), so I will have to close and re-open this PR from a different branch
